### PR TITLE
Assignment 2 - Spencer West -  Matthew Powers  

### PR DESCRIPTION
**Pull request by SaaWest**
turned in

### CODE REVIEW

**Comment by colestewart12** [comment]
I can review this!

File: Logger/FileLogger.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    public string filePath;
+
+    public FileLogger(string FilePath)
+    {
+        this.filePath = FilePath;
+    }
+
+    public override void Log(LogLevel logLevel, string message)
+    {
+
```
Comment (colestewart12) [inline Logger/FileLogger.cs:17]: Hint: Can implement using the File.AppendAllText() with the filePath as the first parameter and the string output you want to produce as the second(date + class + level + message).

File: Logger/LogFactory.cs [hunk 2]
```diff
@@ -2,9 +2,25 @@
...
+        {
+            return null;
+        }
     }
+
+    public string? ConfigureFileLogger(string? filePath)
+    {
+        File = filePath;
+        return File;
+    }
```
Comment (colestewart12) [inline Logger/LogFactory.cs:23]: Make sure to write test for this method!

**Review by colestewart12 — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ❌
The name of the class that created the logger ❌
The log level ❌
The message ❌
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ❌
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested(still working on this it looks like).

Overall, a good start but still needs some implementation of both the tests and classes.
Tests are passing for the most part locally but the build is still failing on the pr during the .Net build stage because of an incorrect `global.json`.

File: Logger/FileLogger.cs (hunk 1 above)
Comment (SaaWest) [reply Logger/FileLogger.cs:17]: I will try that out

File: Logger/BaseLoggerMixinsHelpers.cs
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger/BaseLoggerMixinsHelpers.cs:1]: cannot see what you are referring to specifically

File: Logger.Tests/FileLoggerTests.cs
[the remark this answers is not in the document]
Comment (SaaWest) [reply Logger.Tests/FileLoggerTests.cs:1]: working on it

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 3]
```diff
@@ -14,26 +14,82 @@ public void Error_WithNullLogger_ThrowsException()
         // Arrange
 
         // Act
-        //BaseLoggerMixins.Error(null, "");
+        BaseLoggerMixins.Error(null, "");
+        throw new ArgumentNullException();
```
[the remark this answers is not in the document]
Comment (Matthew-Powers) [reply Logger.Tests/BaseLoggerMixinsTests.cs:18]: Fixed

File: Logger.Tests/LogFactoryTests.cs [hunk 4]
```diff
@@ -6,5 +6,29 @@ namespace Logger.Tests;
...
+        Assert.IsNotNull(fileLog);
+    }
+
+    [TestMethod]
+    [DataRow("file")]
+    public void TryCreateNewLogger_NoClass_Fail(string fileName, string className)
+    {
+        LogFactory logFactory = new();
+        FileLogger? fileLog = logFactory.CreateLogger(fileName, className);
+        //Assert.Fail(fileLog.ClassName != null);
```
[the remark this answers is not in the document]
Comment (Matthew-Powers) [reply Logger.Tests/LogFactoryTests.cs:24]: Fixed

File: Logger.Tests/LogFactoryTests.cs [hunk 5]
```diff
@@ -6,5 +6,29 @@ namespace Logger.Tests;
...
+    [DataRow("file", "class")]
+    public void TryCreateNewFileLogger_NotNull_Success(string fileName, string className)
+    {
+        LogFactory logFactory = new();
+        FileLogger? fileLog = logFactory.CreateLogger(fileName, className);
+        Assert.IsNotNull(fileLog);
+    }
+
+    [TestMethod]
+    [DataRow("file")]
```
[the remark this answers is not in the document]
Comment (Matthew-Powers) [reply Logger.Tests/LogFactoryTests.cs:19]: Fixed

File: Logger/LogFactory.cs (hunk 2 above)
Comment (Matthew-Powers) [reply Logger/LogFactory.cs]: Fixed

**Comment by StrangeRanger** [comment]
@colestewart12 Make sure that all requirements that have been marked with an '❌', have their appropriate code commented on.

Example:

Use the nameof() operator when identifying the class name to the logger ❌

If you thought that the above requirement was not complete, make sure to comment on the code where nameof() should have been used.

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply Logger.Tests/BaseLoggerMixinsTests.cs:17]: Doesn't look like it got fixed

File: Logger.Tests/Logger.Tests.csproj [hunk 6]
```diff
@@ -1,8 +1,14 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
+    <TargetFramework>net8.0</TargetFramework>
     <IsPackable>false</IsPackable>
+	  <LangVersion>12.0</LangVersion>
+	  <Nullable>enable</Nullable>
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:7]:
```suggestion
	  <TargetFramework>net8.0</TargetFramework>
	  <IsPackable>false</IsPackable>
	  <LangVersion>12.0</LangVersion>
	  <Nullable>enable</Nullable>
```

File: Logger/BaseLoggerMixins.cs [hunk 7]
```diff
@@ -1,6 +1,60 @@
-﻿namespace Logger;
+﻿//baseloggerMixins
+using System;
+using System.Globalization;
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
-    
+    public static void Error(this BaseLogger? logger, string message, params string[] arguement)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:8]: Why is logger nullable? I expect to call this message, a logger is needed.

**Pull request closed without merging**